c/meterpreter/source/extensions/powershell/powershell_bridge.cpp

```diff
@@ -524,15 +524,15 @@ DWORD powershell_channel_write(Channel* channel, Packet* request, LPVOID context
 
 	_bstr_t output;
 
-	DWORD result = invoke_ps_command(shell->session_id, codeMarshall, output);
-	if (result == ERROR_SUCCESS && shell->wait_handle)
+	HRESULT result = invoke_ps_command(shell->session_id, codeMarshall, output);
+	if (result == S_OK && shell->wait_handle)
 	{
 		met_api->lock.acquire(shell->buffer_lock);
 		shell->output += output;
 		SetEvent(shell->wait_handle);
 		met_api->lock.release(shell->buffer_lock);
 	}
-	return result;
+	return (DWORD)result;

```

Comment:
This seems problematic since result is an HRESULT but it's being returned as a DWORD where it's then assumed to be a Win32 error. https://github.com/rapid7/metasploit-payloads/blob/master/c/meterpreter/source/metsrv/base_dispatch.c#L901

It would probably be better to always return `ERROR_SUCCESS` when `SUCCEEDED(result)` and a proper Win32 error code otherwise. That may be the trick though, because it doesn't look like you can easily convert an HRESUT back to a Win32 error code. If it'a always accessible from `GetLastError` that could work, otherwise, something generic would be a good idea (maybe `ERROR_FUNCTION_FAILED`?).

Reply:
Yup that was one of the main issues I ran into, not being able to convert HRESULT into Win32 easily :grimacing: 
I'm happy to return `ERROR_FUNCTION_FAILED` though I think that makes sense